store/tikv/coprocessor.go

```diff
@@ -456,84 +465,158 @@ func (it *copIterator) Next() ([]byte, error) {
+	remainTasks := []*copTask{task}
+	for len(remainTasks) > 0 {
+		tasks, err := it.handleTaskOnce(goCtx, bo, remainTasks[0], ch)
+		if err != nil {
+			ch <- copResponse{err: err}
+			return
+		}
+		if len(tasks) >= 0 {
+			remainTasks = append(tasks, remainTasks[1:]...)
+		} else {
+			remainTasks = remainTasks[1:]
+		}
+	}
+}
+
+// handleTaskOnce handles single copTask, successful results are send to channel.
+// If error happened, returns error. If region split or meet lock, returns the remain tasks.
+func (it *copIterator) handleTaskOnce(goCtx goctx.Context, bo *Backoffer, task *copTask, ch chan copResponse) ([]*copTask, error) {
 	sender := NewRegionRequestSender(it.store.regionCache, it.store.client)
+	task.storeAddr = sender.storeAddr
+	req := &tikvrpc.Request{
+		Type: task.cmdType,
+		Cop: &coprocessor.Request{
+			Tp:     it.req.Tp,
+			Data:   it.req.Data,
+			Ranges: task.ranges.toPBRanges(),
+		},
+		Context: kvrpcpb.Context{
+			IsolationLevel: pbIsolationLevel(it.req.IsolationLevel),
+			Priority:       kvPriorityToCommandPri(it.req.Priority),
+			NotFillCache:   it.req.NotFillCache,
+		},
+	}
+	resp, err := sender.SendReq(bo, req, task.region, ReadTimeoutMedium)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if task.cmdType == tikvrpc.CmdCopStream {
+		return it.handleCopStreamResult(goCtx, bo, resp.CopStream, task, ch)
+	}
+
+	tasks, err1 := it.checkCopResponse(bo, resp.Cop, task)
+	if err1 != nil {
+		return nil, errors.Trace(err1)
+	}
+	if len(tasks) > 0 {
+		return tasks, nil
+	}
+
+	it.sendToRespCh(goCtx, copResponse{Response: resp.Cop}, ch)
+	return nil, nil
+}
+
+func (it *copIterator) handleCopStreamResult(goCtx goctx.Context, bo *Backoffer, stream tikvpb.Tikv_CoprocessorStreamClient, task *copTask, ch chan copResponse) ([]*copTask, error) {
 	for {
-		select {
-		case <-it.finished:
-			return nil
-		default:
+		resp, err := stream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, errors.Trace(err)
 		}
 
-		req := &tikvrpc.Request{
-			Type: tikvrpc.CmdCop,
-			Cop: &coprocessor.Request{
-				Tp:     it.req.Tp,
-				Data:   it.req.Data,
-				Ranges: task.ranges.toPBRanges(),
-			},
-			Context: kvrpcpb.Context{
-				IsolationLevel: pbIsolationLevel(it.req.IsolationLevel),
-				Priority:       kvPriorityToCommandPri(it.req.Priority),
-				NotFillCache:   it.req.NotFillCache,
-			},
+		tasks, err := it.checkCopResponse(bo, resp, task)
+		if err != nil || len(tasks) != 0 {
+			return tasks, errors.Trace(err)
 		}
-		resp, err := sender.SendReq(bo, req, task.region, ReadTimeoutMedium)
-		if err != nil {
-			return []copResponse{{err: errors.Trace(err)}}
+
+		it.sendToRespCh(goCtx, copResponse{resp, nil}, ch)
+	}
+	return nil, nil
+}
+
+// checkCopResponse checks coprocessor Response for region split and lock.
+func (it *copIterator) checkCopResponse(bo *Backoffer, resp *coprocessor.Response, task *copTask) ([]*copTask, error) {
```

Comment:
The function name doesn't describe its behavior.
I think it's better to only check the response, and return a bool to indicate if the task needs rebuild, then explicitly rebuild the task outside of the function.